test/clustering_quality.jl

```diff
@@ -0,0 +1,58 @@
+using Test
+using Clustering, Distances
+using OffsetArrays
+
+@testset "clustering_quality()" begin
+
+    # test data with 4 clusters
+
+    Y = [-2 4; 2 4; 2 1; 3 0; 2 -1; 1 0; 2 -4; -2 -4; -2 1; -1 0; -2 -1; -3 0]
+    C = [0 4; 2 0; 0 -4; -2 0]
+    A = [1, 1, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4]
+    W = [
+        1 0 0 0
+        1 0 0 0
+        0 1 0 0
+        0 1 0 0
+        0 1 0 0
+        0 1 0 0
+        0 0 1 0
+        0 0 1 0
+        0 0 0 1
+        0 0 0 1
+        0 0 0 1
+        0 0 0 1
+    ]
+
+    # visualisation of the data
+    # using Plots
+    # scatter(Y[:,1],Y[:,2],
+    #     axisratio = :equal,
+    #     #seriescolor = palette(default)[A],
+    # )
+    # scatter!(C[:,1],C[:,2],
+    #     marker = :square,
+    #     label = "cluster centers",  
+    # )
+
+    @testset "input checks" begin
+        @test_throws ArgumentError clustering_quality(zeros(2,2), zeros(2,3), [1, 2], quality_index = :calinski_harabasz)
+        @test_throws DimensionMismatch clustering_quality(zeros(2,2),zeros(3,2), [1, 2], quality_index = :calinski_harabasz)
+        @test_throws ArgumentError clustering_quality(zeros(2,2),zeros(2,1), [1, ], quality_index = :calinski_harabasz)
+        @test_throws ArgumentError clustering_quality(zeros(2,2),zeros(2,2), [1, 2], quality_index = :calinski_harabasz)
+        @test_throws DimensionMismatch clustering_quality([1,2,3], zeros(2,2), quality_index = :dunn)
+    end
+
```

Comment:
We should also test corner cases to make sure they are handled correctly (ideally, the correct value according to the quality index definition should be returned):
  * degenerated input data: 0 points (and 0 clusters)
  * trivial clustering: single cluster
  * 0-dimensional data

Reply:
Quality indices are generally undefined for single cluster. The reason is that the cluster spread is undefined.

Reply:
Ok, *silhouettes()* actually throws an error for the degenerated clustering, so the other indices should follow that behaviour.

The other corner case that *silhouettes()* checks are empty clusters (gaps in `assignments` values).
See [test/silhouette.jl](https://github.com/JuliaStats/Clustering.jl/blob/master/test/silhouette.jl).

Reply:
If the cluster is empty than it's center is undefined, which means that the user gave nonsense arguments (or inconsistent number of assignments/centers, this is checked). Only silhouettes and Dunn index do not require them.

Reply:
ideally we need to test that the behavior in such cases meets our expectations, because this will happen (bugs in the user code). I'd say: let's add the tests, see how `clustering_quality` handles it  and then decide if we need to adjust the behavior.